### PR TITLE
fix: recompile&register scripts as UDF on reboot

### PR DESCRIPTION
**Pull request by discord9**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Recompile and re-register scripts on reboot

register scripts in create `scripts` table's callback
currently insert new row into `scripts` table wouldn't trigger compile and register of new python script, so sqlness tests can't be done yet, might impl it later

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

### CODE REVIEW

File: src/catalog/src/error.rs [hunk 1]
```diff
@@ -27,6 +27,11 @@ use crate::DeregisterTableRequest;
 #[derive(Debug, Snafu)]
 #[snafu(visibility(pub))]
 pub enum Error {
+    #[snafu(display("Failed to re-compile script scripts table, source: {}", source))]
+    CompileScript {
+        #[snafu(backtrace)]
```
Comment (killme2008) [inline src/catalog/src/error.rs:32]: Adds the script name to error message.

File: src/script/src/manager.rs [hunk 2]
```diff
@@ -38,20 +38,17 @@ impl ScriptManager {
         catalog_manager: CatalogManagerRef,
         query_engine: QueryEngineRef,
     ) -> Result<Self> {
-        Ok(Self {
+        let zelf = Self {
```
Comment (killme2008) [inline src/script/src/manager.rs:41]: Why does this change? Looks unnecessary.

File: src/script/src/manager.rs [hunk 3]
```diff
@@ -65,6 +62,18 @@ impl ScriptManager {
         Ok(script)
     }
 
+    /// compile script to PyScript, but not register them to the query engine and UDF registry nor caching in `compiled`
+    async fn compile_wout_cache(
```
Comment (killme2008) [inline src/script/src/manager.rs:66]: wout -> without?

File: src/script/src/python/engine.rs [hunk 4]
```diff
@@ -179,6 +179,14 @@ pub struct PyScript {
...
+        let copr = Arc::new(parse::parse_and_compile_copr(
+            script,
+            Some(query_engine.clone()),
+        )?);
+
+        Ok(PyScript { copr, query_engine })
+    }
     /// Register Current Script as UDF, register name is same as script name
     /// FIXME(discord9): possible inject attack?
     pub fn register_udf(&self) {
```
Comment (killme2008) [inline src/script/src/python/engine.rs:192]: maybe we can change this function to async and spawn task outside? And `join_all` futures for compiling all scripts.

File: src/script/src/table.rs [hunk 5]
```diff
@@ -48,6 +55,73 @@ pub struct ScriptsTable {
...
+            .context(CompileScriptSnafu)?;
+        let records = record_util::collect(rbs)
+            .await
+            .map_err(BoxedError::new)
+            .context(CompileScriptSnafu)?;
+        if records.is_empty() {
+            // scripts table is empty, no need to recompile
+            return Ok(());
+        }
+        assert_eq!(records.len(), 1);
```
Comment (killme2008) [inline src/script/src/table.rs:81]: We can't ensure that there is only one batch. Let's iterate the batches and do the below stuff.

File: src/script/src/table.rs [hunk 6]
```diff
@@ -48,6 +55,73 @@ pub struct ScriptsTable {
...
+        let record = &records[0];
+        let get_col_by_name = |name: &str| -> Result<_> {
```
Comment (killme2008) [inline src/script/src/table.rs:83]: I don't really like such an inline function, I prefer to extract it as a small function in current mode or struct.

File: src/script/src/table.rs [hunk 7]
```diff
@@ -48,6 +55,73 @@ pub struct ScriptsTable {
...
+            Ok(column)
+        };
+        let names = get_col_by_name("name")
+            .map_err(BoxedError::new)
+            .context(CompileScriptSnafu)?;
+        let scripts = get_col_by_name("script")
+            .map_err(BoxedError::new)
+            .context(CompileScriptSnafu)?;
+        let mut script_list: Vec<(String, String)> = Vec::with_capacity(names.len());
+        for i in 0..names.len() {
```
Comment (killme2008) [inline src/script/src/table.rs:105]: what about using zip and collect? The code would be cleaner.

File: src/script/src/table.rs [hunk 8]
```diff
@@ -48,6 +55,73 @@ pub struct ScriptsTable {
...
+            script_list.push((
+                names.get_data(i).unwrap().to_string(),
+                scripts.get_data(i).unwrap().to_string(),
+            ));
+        }
+        for (name, script) in script_list {
+            if let Ok(script) = PyScript::from_script(&script, query_engine.clone()) {
+                script.register_udf();
+
+                logging::info!(
```
Comment (killme2008) [inline src/script/src/table.rs:115]: debug level is enough.

File: src/script/src/table.rs [hunk 9]
```diff
@@ -201,6 +283,57 @@ impl ScriptsTable {
         Ok(script_column.get_data(0).unwrap().to_string())
     }
 
+    /// simple query and return all scripts in scripts table in the form of `(name, script)`.
+    pub async fn find_all_scripts(&self) -> Result<Vec<(String, String)>> {
```
Comment (killme2008) [inline src/script/src/table.rs:287]: Looks it's not used at all.

File: src/script/src/table.rs [hunk 10]
```diff
@@ -48,6 +56,98 @@ pub struct ScriptsTable {
...
+                    Err(err) => {
+                        logging::warn!(
+                            r#"Failed to compile script "{name}"" in `scripts` table: {err}"#
+                        );
+                        None
+                    }
+                }
+            })
+            .collect();
+        while let Some(res) = set.join_next().await {
```
Comment (killme2008) [inline src/script/src/table.rs:141]: why not just `futures::future::try_join_all(handles)`?

File: src/script/src/table.rs [hunk 11]
```diff
@@ -48,6 +56,98 @@ pub struct ScriptsTable {
...
+                                "Script in `scripts` system table re-register as UDF: {}",
+                                name
+                            );
+                        });
+                        Some(handle)
+                    }
+                    Err(err) => {
+                        logging::warn!(
+                            r#"Failed to compile script "{name}"" in `scripts` table: {err}"#
+                        );
```
Comment (waynexia) [inline src/script/src/table.rs:135]: Does this work? I'm not sure how raw string acts in this macro

**Review by waynexia — COMMENTED** [review]
Generally looks good to me

**Review by killme2008 — APPROVED** [review]
LGTM

**Review by waynexia — APPROVED** [review]
LGTM

**Pull request merged**